File: basic/sort/merge_sort.cpp
#include <iostream>
#include <vector>

class Solution {
 public:
  void merge(int arr[], size_t l, size_t m, size_t r) {
    std::vector<int> left(arr + l, arr + m + 1), right(arr + m + 1, arr + r + 1);// 注意m+1, 左闭右开
    size_t i = 0, j = 0;
    for (size_t a = l; a <= r; ++a) {
      if (i < left.size() && j < right.size()) {// 列举各种情况
        arr[a] = left[i] < right[j] ? left[i++] : right[j++]; 
      } else if (i < left.size() && j >= right.size()) {
        arr[a] = left[i++];
      } else {
        arr[a] = right[j++];
      }
    }
  }

  void merge_sort(int arr[], size_t l, size_t r) {
    if (l >= r) {
      return;
    }
    size_t m = (l + r) / 2;
    merge_sort(arr, l, m);
    merge_sort(arr, m + 1, r);
    merge(arr, l, m, r);
  }
};

int main() {
  int a[6] = {1, 3, 5, 2, 5, 7};

  Solution solution = Solution();
  solution.merge_sort(a, 0, 5);

  for (int num : a) {
    std::cout << num << ' ';
  }
  std::cout << std::endl;
}

File: basic/sort/merge_sort.py
class Merge_Sort(object):

    def ____init__(self) -> None:
        super().__init__()

    def merge(self, arr, l, m, r):
        '''合并两顺序数组
        
        @param:
            arr:     list    输入数组
            l:       int     左边界开始索引
            m:       int     右边界开始索引
            r:       int     右边界结束索引+1
        @Note:
            复制两子数组->合并两数组到原数组对应位置
        '''
        # 复制子数组
        a = [x for x in arr[l:m]]
        b = [x for x in arr[m:r]]
        # 合并两顺序数组
        a_idx, b_idx, arr_idx = 0, 0, l
        while True:
            if a[a_idx] < b[b_idx]:
                arr[arr_idx] = a[a_idx]
                a_idx += 1
                arr_idx += 1
            else:
                arr[arr_idx] = b[b_idx]
                b_idx += 1
                arr_idx += 1
            if a_idx == len(a) or b_idx == len(b):  #注意边界条件
                break

        if a_idx != len(a):
            arr[arr_idx:r] = a[a_idx:]
        if b_idx != len(b):
            arr[arr_idx:r] = b[b_idx:]

    def sort(self, arr, l, r):
        '''
        
        @Note:
            左排序->右排序->合并
        '''
        if l >= r:  #递归终止条件
            return
        m = (l + r) // 2
        self.sort(arr, l, m)
        self.sort(arr, m + 1, r)
        self.merge(arr, l, m + 1, r + 1)


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    merge_sort = Merge_Sort()
    merge_sort.sort(arr, 0, len(arr) - 1)
    print(arr)
